src/transformers/trainer.py

```diff
@@ -818,17 +818,43 @@ def create_optimizer(self):
                     "weight_decay": 0.0,
                 },
             ]
-            optimizer_cls = Adafactor if self.args.adafactor else AdamW
-            if self.args.adafactor:
+            if self.args.adafactor and self.args.optim not in {"adamw_hf", "adafactor"}:
+                raise ValueError(f"You passed the --adafactor flag and optimizer {self.args.optim}.")
+
+            optimizer_kwargs = {"lr": self.args.learning_rate}
+
+            adam_kwargs = {
+                "betas": (self.args.adam_beta1, self.args.adam_beta2),
+                "eps": self.args.adam_epsilon,
+            }
+
+            # TODO the following code is a good candidate for PEP 622 once Python 3.10 becomes the
+            #  minimum required version. See, https://www.python.org/dev/peps/pep-0622/
+            if (self.args.adafactor and self.args.optim == "adamw_hf") or self.args.optim == "adafactor":
                 optimizer_cls = Adafactor
-                optimizer_kwargs = {"scale_parameter": False, "relative_step": False}
-            else:
+                optimizer_kwargs.update({"scale_parameter": False, "relative_step": False})
+            elif self.args.optim == "adamw_hf":
+                from .optimization import AdamW
+
                 optimizer_cls = AdamW
-                optimizer_kwargs = {
-                    "betas": (self.args.adam_beta1, self.args.adam_beta2),
-                    "eps": self.args.adam_epsilon,
-                }
-            optimizer_kwargs["lr"] = self.args.learning_rate
+                optimizer_kwargs.update(adam_kwargs)
+            elif self.args.optim == "adamw_torch":
+                from torch.optim import AdamW
+
+                optimizer_kwargs.update(adam_kwargs)
```

Comment:
The `optimizer_cls` line is missing in this part of the test.

Reply:
Good catch. It should be fixed now.